Boiler Fuel Allocation Improvements

grgmiller opened the pull request:
Not sure if it is appropriate to create a pull request yet, or if I should wait until I've completed more tests, but the PR represents my work on this project over the past week. I don't yet have the pudl dev environment set up, so I haven't run any pre-commit hooks, etc. I can do this in a future PR though. 

Summary of changes:
 - allocate generation and fuel by month, rather than by year (requires that `pudl_out` have a frequency of 'MS' set). This requires making duplicate entries in the `gens` table for each month to help with merging data by `report_date`
 - adds a function `manually_fix_prime_movers` that patches some of the issues identified in #1585 
 - Previously, the code was ignoring any startup fuels that were used, if these differed from any of the six `energy_source_code`. I've added a new function `add_startup_energy_sources_to_esc_columns` that identifies the unique set of all energy source codes and startup fuel codes, and re-assigns them to numbered energy source codes.
 - Stops using the `remove_bf_nulls` function, since this was also removing reported zero values, which actually helps with properly allocating the fuel
 - Updates the `allocate_fuel_by_gen_esc` function to follow a similar process to the `allocate_net_gen_by_gen_esc` function, except that it looks in the bf table instead of the gf table to calculate fractions
 - Allocates not only `fuel_consumed_mmbtu` but `fuel_consumed_for_electricity_mmbtu`
 - Updates the calculation of `frac_fuel` so that 0 / 0 = 1, instead of NaN, which was causing issues with the allocation
 - Changes `remove_retired_generators` to `remove_inactive_generators`. This function now: keeps all "existing" generators, keeps any "retired" generators whose monthly report_date <= the retirement date, keeps any "proposed" generators starting in the month when they start reporting generation data.
 - Allows the user to specify the `sum_cols` to use in `aggregate_gen_fuel_by_generator` in case they add other custom data columns to the allocated data after running `allocate_gen_fuel_by_generator_energy_source` but before they plug the resulting `net_gen_fuel_alloc` df into `aggregate_gen_fuel_by_generator`. For example, in my case, I calculate a `co2_mass_tons` column based on the fuel type and fuel consumption, which I want to aggregate up to the gen level.

Next Steps:
 - I've done some spot checks on a few plants and the allocation seems reasonable, but I need to do some more in-depth checks: for example when there is negative net generation, or in cases where the frac column does not sum to 1.
 - The `manually_fix_prime_movers` function is a patch, but we should figure out how to address the issue of prime mover codes not matching up, since that impacts the accuracy of the allocation

grgmiller commented:
This newest commit seems to fix the warnings generated by `_test_frac()`, where the fraction doesn't add up to 1. 

This involves two primary fixes:
1. Instead of fixing the divide by zero issue by manually assigning 0/0 = 1, I instead replace all zeros in the columns suffixed by `_g_tbl` or `_gf_tbl` or `_bf_tbl` with a very small number (0.001). This way, if there are multiple generators in a group reporting zeros, the frac still adds to one.
2. matches the data from the `bf` table using `IDX_PM_ESC` instead of `IDX_U_ESC`

I'm still doing some checks on the final values, but this at least passes the test_frac tests.

grgmiller commented:
Commit "fix remove_inactive generators" does the following:
 - fixes an issue where the `remove_inactive_generators` function was incorrectly removing some retiring and proposed generators
 - adds a function `manually_fix_energy_source_codes`, which for now updates the ESC of a single plant (10204), which has a fuel code of OTH in the gf table, but a fuel code of WH in the gens table. In the future, we may want a more programmatic way to fix these mismatched fuel codes.
 - Fixes the doctest in `analysis.timeseries_cleaning` that was causing the pre-commit hooks to fail.

grgmiller commented:
Previously this code had grouped MSN and MSB fuel codes into MSW, since the EIA-860 generator file only uses the MSW code, while the EIA-923 files (in recent years) started splitting this into the biogenic (MSB) and nonbiogenic (MSN) components. However, MSN and MSB have different emissions characteristics, so I wanted to keep these codes separate.

Thus, instead of solving this key issue by grouping MSN and MSB into MSW, this new code does the opposite: if EIA-923 uses the MSN and MSB codes, the code now splits the MSW code in EIA-860 into two separate energy source codes (MSB and MSN) so that the data from each source can successfully be merged, while preserving the different characteristics of these two fuels.

grgmiller commented:
I made some updates to how plants with proposed and retired operating statuses were being handled. 

Previously, for proposed generators, we were only keeping the records if they reported generator-specific data from the generation table. We were excluding proposed generators with non-zero generation data from the gf table, because there was a chance that this gf data was being allocated based on an existing generator at the same plant that shared a prime mover. 

However, if a proposed generator is part of an existing plant, it is possible that this gf generation belongs to one of the other existing generators. Thus, we want to identify those proposed generators where the entire plant is proposed (since the gf-reported generation could only come from one of the new generators).

We take a similar approach for retired plants, where we want to make sure that we keep data for retired plants that have nonzero gf generation data only if the entire plant had been marked as retired prior to the previous year. Otherwise, there is a chance that the gf table data that was allocated to a generator was from an existing generator, or a generator that is retiring in the current year.

grgmiller commented:
hi @cmgosnell and @zaneselvans  - I think this PR is ready for your review. It looks like I need your approval to run the full set of checks on the PR. 

A few notes/questions:
 - We might want to change the name of the module since this now allocates fuel and generation, and not just net generation. I might suggest something more generic like `allocate_eia923.py`
 - there are two functions in the module that perform manual fixes to the underlying data that is the result of several known issues. These functions have only been fully tested using year 2020 data. Let me know how you think we should handle these and if these have already been fixed elsewhere. 
     - `manually_fix_prime_movers()` manually updates prime mover codes that are incorrect until https://github.com/catalyst-cooperative/pudl/issues/1585 is resolved. It looks like maybe this has been fixed in the dev branch?
      - `manually_fix_energy_source_codes()` updates the `energy_source_code` for a single plant (10204) in the `gf` table from `OTH` to `WH` in order to match the fuel code in the `gens` table

grgmiller commented:
Thanks Zane! The code was passing all of the pre-commit hooks, so if `tox` is included in there, then yes. However, I haven't run tox separately. I can look into the docs on how to do that if that's a separate process from the pre-commit hooks.

I'll wait for you to merge `dev` in and I can try re-running the tests locally. 

#1600 might have fixed the prime-mover issue. I've been using the updated `pudl.sqlite` file downloaded from datasette for testing, so I'll look into whether this function is actually still necessary. I'll also look at #1585 and respond to that comment.

zaneselvans commented:
I think the ideal place to apply that fix to the non-overlapping energy source codes would probably be in the original ETL. Would it be hard to compile a table of all the cases in which these codes don't line up in the different tables?

grgmiller commented:
So now that I've merged in the updated `allocate_fuel` branch, I'm getting the following error when I try and update the `pudl-dev` environment:

```
Installing pip dependencies: / Ran pip subprocess with arguments:
['A:\\miniconda3\\envs\\pudl-dev\\python.exe', '-m', 'pip', 'install', '-U', '-r', 'A:\\GitHub\\pudl\\devtools\\condaenv.aawuhsrv.requirements.txt']
Pip subprocess output:
Obtaining file:///A:/GitHub/pudl (from -r A:\GitHub\pudl\devtools\condaenv.aawuhsrv.requirements.txt (line 1))
  Installing build dependencies: started
  Installing build dependencies: finished with status 'error'

Pip subprocess error:
  error: subprocess-exited-with-error

  × pip subprocess to install build dependencies did not run successfully.
  │ exit code: 1
  ╰─> [18 lines of output]
      Collecting setuptools<63
        Using cached setuptools-62.6.0-py3-none-any.whl (1.2 MB)
      Collecting pip<23,>=21
        Using cached pip-22.1.2-py3-none-any.whl (2.1 MB)
      Collecting wheel
        Using cached wheel-0.37.1-py2.py3-none-any.whl (35 kB)
      Collecting setuptools_scm
        Using cached setuptools_scm-7.0.4-py3-none-any.whl (42 kB)
      Collecting typing-extensions
        Using cached typing_extensions-4.3.0-py3-none-any.whl (25 kB)
      Collecting packaging>=20.0
        Using cached packaging-21.3-py3-none-any.whl (40 kB)
      Collecting tomli>=1.0.0
        Using cached tomli-2.0.1-py3-none-any.whl (12 kB)
      Collecting pyparsing!=3.0.5,>=2.0.2
        Using cached pyparsing-3.0.9-py3-none-any.whl (98 kB)
      ERROR: To modify pip, please run the following command:
      A:\miniconda3\envs\pudl-dev\python.exe -m pip install --ignore-installed --no-user --prefix C:\Users\Greg\AppData\Local\Temp\pip-build-env-a7pe0_wj\overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- setuptools<63 pip>=21,<23 wheel setuptools_scm
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

failed

CondaEnvException: Pip failed
```

For now, I may revert this branch to https://github.com/catalyst-cooperative/pudl/pull/1608/commits/a49e2013e981884b225d4bb1d95fad848a987690 so that the code is still usable

grgmiller commented:
So I tried resetting my local repo to the previous commit and creating the pudl-dev environment, but now I am still getting the same error as above. 

It seems to come at the last step when running ` - --editable ../[doc,test,dev,datasette]` in the environment file.

grgmiller commented:
Ok so the result of all of this is that I now have a working `pudl-dev` environment, but I am unable to run `tox` locally due to an issue with fiona / GDAL on Windows. 

However, based on the online tox workflow, it appears that there are only 2 issues:
 - something about a slack webhook: `Error: Specify secrets.SLACK_WEBHOOK_URL`
 - The failing doc8 issue due to pip in their pyproject file.

zaneselvans commented:
The `secrets.SLACK_WEBHOOK_URL` bit doesn't work because it's an authentication token that's only available within the Catalyst organization, and GitHub can tell that you're outside the org. So that's not something to worry about.

That's too bad about the GDAL / fiona issue. You should still be able to run `pytest` directly within the `pudl-dev` environment and have it work (minus the environment isolation that Tox provides).  I think if you just run `pytest` with no arguments it will run a substantial portion of the tests that would have been run by `tox`, and all of the CI will get run when you push to the branch (though the slack notification will fail because of the secrets permissions).

Right now it looks like there's [a failing doctest](https://github.com/catalyst-cooperative/pudl/runs/7260371526?check_suite_focus=true#step:9:962) in the `timeseries_cleaning` module.

grgmiller commented:
Strange - previously that doctest was failing because the dtype was not specified, and now it's failing because the dtype is specified! I'm wondering if for some reason the doctest in the pre-commit virtual environment failing differently than the doctest in tox. That doctest passed the pre-commit tests

grgmiller commented:
So when I run pytest, I get the following errors:
```
FAILED test/integration/output_test.py::test_eia_outputs[gens_eia860-frc_eia923-12.0-kwargs7] - requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
FAILED test/integration/output_test.py::test_eia_outputs[gens_eia860-gen_fuel_by_generator_eia923-1.0-kwargs9] - AssertionError: Expected ratio of distinct report_date values to be 1.0, but found 24 / 2 = 12.0
FAILED test/integration/output_test.py::test_eia_outputs[gens_eia860-fuel_cost-12.0-kwargs15] - requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
FAILED test/integration/output_test.py::test_eia_outputs[gens_eia860-mcoe-12.0-kwargs17] - requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
FAILED test/integration/output_test.py::test_null_rows[mcoe-0.9] - requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
FAILED test/integration/output_test.py::test_ferc714_outputs[summarize_demand] - FileNotFoundError: [WinError 2] The system cannot find the file specified
FAILED test/integration/output_test.py::test_ferc714_respondents_georef_counties - FileNotFoundError: [WinError 2] The system cannot find the file specified
FAILED test/integration/output_test.py::test_plant_parts_eia_filled - requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
FAILED test/integration/output_test.py::test_mcoe_filled[mcoe-0.2-0.02] - requests.exceptions.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```
It looks like most (or all) of these are unrelated to the `allocate_net_gen` module and primarily seems to be an issue with the EIA API - it looks like the API is returning a file not found error what whatever API calls it is making.

zaneselvans commented:
Do you have an `API_KEY_EIA` environment variable set so it can access the fuel price estimates from EIA?

If you do have an API key set then  yeah the `JSONDecodeError`s are probably just the API being flaky (as it is ~half the time, which is part of why we're trying to replace it in #1708).

I think the `gen_fuel_by_generator` error might be because the tests previously expected that generation (and fuel) allocation would only work on an annual basis, so even in the monthly case that you've now enabled they're looking for the expected number of annual records, when they need to be expecting to see monthly aggregations.

I suspect the `ferc714` test errors where it's not finding a file are due to the fact that `ogr2ogr` can't easily be installed on Windows, and it is needed to convert the US Census DP1 geodatabase into an SQLite DB for use in the balancing area calculations. This was one of several issues that eventually led me to give up on getting the whole system working under Windows.

grgmiller commented:
I did specify an `API_KEY_EIA` per the instructions [here](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/dev_setup.html#getting-and-storing-an-eia-api-key) (as a side note, these instructions specify putting single quotes around the API key, but that did not work for me - I had to just enter the key without quotes after the equal sign). 

I can update the `gen_fuel_by_generator` test to expect 12 values.

zaneselvans commented:
Hmm, I imagine the API key quotes thing is also a Windows quirk. Not sure.

I would say if you have to choose between making `tox` tests and `pre-commit` checks work because of the Windows environment issues, opt for making `tox` work since it's a much more complete set of tests, and it's what will be run via GitHub actions in CI.  If your `git commit` fails due to a hook that doesn't pass, and which you expect not to pass as a known Windows problem, you can re-run the commit using `git commit --no-verify` and it will bypass the hooks.

I'm really not sure what the right thing to do is on the energy source code mismatches. Making a bunch of tweaks to the contents of the tables far downstream from the database seems like a bad solution, since it'll result in inconsistencies between the data that's being stored as the single source of truth (the normalized DB tables) and the analytical outputs.  Allowing every user to apply their own different fixes also sounds like a recipe for confusion.

How many tables & records do these inconsistencies crop up? Can fixes be applied such that the `generators_eia860`, `generation_fuel_eia923` and `boiler_fuel_eia923 tables are all consistent with each other during the initial ETL process so what's in the DB is internally consistent?  Can we have a single per-record fix-applying process that's applied to all of those tables during the transform process so we know they're all getting the same treatment?

I don't think we have to make this decision to get your PR merged into the `allocate_fuel` branch, but I'd like to get @cmgosnell to look over all of this and offer feedback before we merge it into `dev`. She's out this week biking in the Colorado Rockies though.

grgmiller commented:
In the 2020 data, I've only noticed this for a single plant/fuel combo (so 12 records?) in the `gf_eia923` table. 

This really only seems to be an issue with OTH energy source codes. One of the things that I've been working on in the eGRID code is replacing all instances of OTH with an actual fuel code. This involves manually identifying what the most likely fuel code is for fuels that are reported as OTH, based on 1) the reported heat content of the fuel and comparing it to the heat content of other labeled fuels, and 2) examining the plant name for clues (ie if the plant is part of a refinery or a phosphate plant, it can help identify the fuel used.

You probably don't want to implement these changes in the ETL process if your goal is to preserve the reported source data though. 

In this specific case for plant `10204`, it seems like EIA-860 consistently reports `WH` as a fuel, while EIA-923 consistently reports `OTH` as a fuel. It is possible that these are actually two separate fuels, but I find it unlikely.

Perhaps a more programmatic way to update this in `allocate_net_gen` without manually updating the fuel code would be to:
1. Identify primary keys in `gf_eia923` and `gens_eia860` that don't have a match in the other table
2. For these mismatched primary keys, if one of them has an energy source code of OTH, update it to match the fuel code of the mismatched key from the other table

zaneselvans commented:
The medium-term plan is to write the tables into a data warehouse at several significant steps along their transformation journey, so we'd have a relatively untouched version of the table available prior to this kind of cleaning, and later version with this kind of internal inconsistency fixed. This will provide several different potential starting points for folks depending on what kind of changes they're trying to make or how refined / raw an input they want to start with (and also make it easier to see where the changes we've applied are taking place).

Is there really sufficient heat content in these `OTH` records to justify the manual cleaning? That seems like it could be crazy tedious.

grgmiller commented:
At least in 2020, there is about 16 million mmbtu of fuel that is assigned an OTH fuel code from 11 different plants. In the grand scheme of all plants nationally, perhaps that is not a lot, but these other fuel types are often pretty dirty fuels like DFO, MSB, and RC, so we didn't want to be missing these emissions in our emissions database. There are many more plants with OTH fuel types that report no fuel or generation data, so we didn't bother filling these in. 

I think that if we actually wanted to programmatically fill these OTH fuel types, there would be a way to do that by looking for keywords in the plant name and also by assigning a fuel based on the reported fuel heat content and the distribution of heat contents reported for other fuels.

grgmiller commented:
When I was double-checking this code, I realized that after loading the `bf_eia923 table` I am dropping the `boiler_id` column so the allocation currently doesn't take the boiler-generator association into account! This is a pretty big issue so I need to spend some time fixing this.

zaneselvans commented:
Oops! Okay. Christina isn't back until next week anyway.

grgmiller commented:
Ok just pushed an update that fixes the boiler fuel allocation issue. 

This update does two primary things:
 - uses the `boiler_generator_assn_eia860` table to correctly associate data in `bf` with the rest of the data. This is coordinated by `allocate_bf_data_to_gens()`. Because fuel consumption in the boiler_fuel_eia923 table is reported per boiler_id, we must first map this data to generators using the boiler_generator_assn_eia860 table. For boilers that have a 1:m or m:m relationship with generators, we allocate the reported fuel to each associated generator based on the nameplate capacity of each generator. So if boiler "1" was associated with generator A (25 MW) and generator B (75 MW), 25% of the fuel consumption would be allocated to generator A and 75% would be allocated to generator B.
 - Certain plants only report data to the generator table and boiler fuel table on an annual basis. In this case, their annual total is reported as a single value in December, and the other 11 months are reported as missing values. Previously, this annually-reported data was only being used to allocate generation and fuel data in December, instead of for the entire year. The raw EIA-923 table includes a column for 'Respondent Frequency' (which includes a code of 'A' for annual and 'M' for monthly). In theory this column could be used to identify these plants, but but this column is not currently included in pudl. Thus, we use the function `distribute_annually_reported_data_to_months()` to do two things. 
     - First, the function programmatically identifies which plants are annual frequency reporters based on the pattern of the data reported. Specifically, it looks for generators or boilers where months 1-11 are NA, and month 12 is not NA.
     - Then, the function distributes this annually reported data evenly to all 12 months so that it can be used to allocate generation and fuel data.

zaneselvans commented:
> Certain plants only report data to the generator table and boiler fuel table on an annual basis. In this case, their annual total is reported as a single value in December, and the other 11 months are reported as missing values. Previously, this annually-reported data was only being used to allocate generation and fuel data in December, instead of for the entire year. The raw EIA-923 table includes a column for 'Respondent Frequency' (which includes a code of 'A' for annual and 'M' for monthly). In theory this column could be used to identify these plants, but but this column is not currently included in pudl. 

Whaaat? Well we should add that. Where does it show up in the spreadsheets? What entity is it associated with (generator, plant, utility?)

That's wild that there's some December-only records for the whole year. How annoying. I wonder what all other tables have this dynamic going on.

zaneselvans commented:
I just merged the most recent changes from `dev` into the `allocate_fuel` branch. It looks like the only CI failure in the last run was some EIA API flakiness, so hopefully we can get this merged in the morning.

grgmiller commented:
Awesome! Just to confirm, do you need me to do anything else at this point?

zaneselvans commented:
Hey @grgmiller since the CI here was failing due to some EIA API flakiness or lack of access to the API_KEY_EIA stored in the repo secrets, I pulled your branch and ran the CI locally.  I'm seeing one thing that seems to be a real failure, which you should be able to reproduce with the following command:

```
pytest test/integration/output_test.py::test_plant_parts_eia_filled
```

I'm getting a failed assertion in the plant parts EIA process, where it finds an NA value for the EIA record ID.
```
src/pudl/analysis/plant_parts_eia.py:1573: AssertionError
```

There's also a separate warning (I have no idea if it's related):
```
test/integration/output_test.py::test_plant_parts_eia_filled
  /home/zane/code/catalyst/pudl/src/pudl/analysis/allocate_net_gen.py:224: UserWarning: 1 generators are missing prime mover codes in gens_eia860. This will result in incorrect allocation.
```

grgmiller commented:
I'm not super familiar with the PPE so maybe it would just be faster for @cmgosnell  to take a look when she's back

grgmiller commented:
Hi @zaneselvans I just noticed that the allocation is leading to a few plants having negative fuel consumption - I'm working on a fix and will push an update soon.

zaneselvans commented:
I merged @katie-lamb's fix to the PPE generation into this branch locally (it's been merged into the `allocate_fuel` branch) and re-rant the failing test, and it's producing the same failure, so that doesn't seem to be the source of the problem.

zaneselvans commented:
Hey @grgmiller, @cmgosnell's train has been extremely delayed so it might be tomorrow or Friday before she's able to focus on this review.

grgmiller commented:
So I've been looking into the single generator in 2020 that raises the warning about a missing prime mover. This is plant 59012 with generator_id "STEAM" (this is the DC Water CHP plant). It looks like in 2015-2018, this generator had data reported in EIA-860, and was listed as a "CA" prime mover, and the other three generators were listed as "CT".

However, it looks like starting in 2019, this generator was removed from EIA-860, and the other three generators changed their prime mover to "GT". This is likely related to the conversation in https://github.com/catalyst-cooperative/pudl/issues/1585. 

So the question is how do we want to handle this? Should this generator still exist with a 2020 record in `gens_eia860`? If so, should it be marked as retired instead of just having all missing data associated with it?

cmgosnell commented on src/pudl/analysis/allocate_net_gen.py:
```diff
@@ -731,6 +1028,121 @@ def allocate_net_gen_by_gen_esc(gen_pm_fuel):
+        # fraction of fuel that does not show up in the bf table
+        # set minimum fraction to zero so we don't get negative fuel
+        frac_missing_from_bf_tbl=lambda x: np.where(
+            (x.frac_from_bf_tbl < 1), (1 - x.frac_from_bf_tbl), 0
```
I really like this solution for avoiding negative allocations!

cmgosnell commented on src/pudl/analysis/allocate_net_gen.py:
```diff
@@ -731,6 +1028,121 @@ def allocate_net_gen_by_gen_esc(gen_pm_fuel):
+        # should be 0 for any generator that does not have fuel in the bf_tbl
+        # and frac_cap should be 0 for any generator that has fuel in the
+        # bf_tbl.
+        frac=lambda x: np.where(x.in_bf_tbl, x.frac_bf, x.frac_cap),
```
even after writing the analogous version of this for net gen, this still feels complicated but I think i've wrapped my head around it and this looks good! I think it would be good to add baby unit tests to demonstrate outcomes for this. If you want I could add some for the net get side of things (which i should have done originally!) as an example/template.

cmgosnell commented on src/pudl/analysis/allocate_net_gen.py:
```diff
@@ -731,6 +1028,121 @@ def allocate_net_gen_by_gen_esc(gen_pm_fuel):
+    # squish all of these methods back together.
+    fuel_alloc = pd.concat([all_gen, some_gen, bf_only])
+    # null out the inf's
+    fuel_alloc.loc[abs(fuel_alloc.frac) == np.inf] = np.NaN
```
re: my 0's/nulls comment before I think I like this solution in general, but if we are going to use the small numbers method to deal with zeros/nulls do we need this at all?

cmgosnell commented on src/pudl/analysis/allocate_net_gen.py:
```diff
@@ -731,6 +1028,121 @@ def allocate_net_gen_by_gen_esc(gen_pm_fuel):
+            # we could x.fuel_consumed_mmbtu_bf_tbl.fillna here if we wanted to
+            # take the net gen
+            fuel_consumed_mmbtu=lambda x: x.fuel_consumed_mmbtu_gf_tbl * x.frac,
+            fuel_consumed_for_electricity_mmbtu=lambda x: x.fuel_consumed_for_electricity_mmbtu_gf_tbl
```
I'm vv glad you added `fuel_consumed_for_electricity_mmbtu`!

Do you think there is any difference between these two fuel columns that should lead to us treating them differently? (for the record I very much hope the answer to this question is "NO IT SHOULD BE THE SAME" for simplicity)

grgmiller commented on src/pudl/analysis/allocate_net_gen.py:
```diff
@@ -158,19 +183,87 @@ def allocate_gen_fuel_by_generator_energy_source(pudl_out, drop_interim_cols=Tru
+            f"{len(missing_pm)} generators are missing prime mover codes in gens_eia860. "
+            "This will result in incorrect allocation."
+        )
+        print(
```
Can do!

grgmiller commented on src/pudl/analysis/allocate_net_gen.py:
```diff
@@ -158,19 +183,87 @@ def allocate_gen_fuel_by_generator_energy_source(pudl_out, drop_interim_cols=Tru
+            ]
+        )
+    # duplicate each entry in the gens table 12 times to create an entry for each month of the year
+    gens = create_monthly_gens_records(gens)
```
I can look into this.

grgmiller replied:
Good point - I haven't tested this updated code with `pudl_out.freq == "AS"`. I'm thinking that perhaps the most effective solution here would be to replace this line with:
```
if pudl_out.freq == "MS":
     gens = pudl.helpers.expand_timeseries(df=gens, key_cols=["plant_id_eia", "generator_id"], freq="MS")
```

I will note however that when testing this, `create_monthly_gens_records()` took 1s to complete, while `expand_timeseries` took 1m37s to complete, so it is quite a bit slower.

grgmiller commented:
@cmgosnell I think that I have responded to all of your specific reviews. Please let me know what follow up you need from me on any of the issues. 

Also, before merging this PR, I'm thinking that we may want to rename the module from `allocate_net_gen` to something like `allocate_generation_fuel_eia923`

grgmiller commented:
I've also been doing some additional digging on the annual-reporter issue and discovered something that we probably want to fix in this PR. As a reminder, plants report to EIA at one of three different frequencies


respondent_frequency | Description
-- | --
A | The respondent only provides an   annual total(s) for this record  via the EIA-923 annual survey form.  Any monthly data in this record is    estimated based on the respondent's reported annual total(s) and power  plants   with similar characteristics to this plant.
AM | The respondent provides monthly values   for this record, but does so once per year via the EIA-923 annual survey   form.
M | The respondent provides monthly values   for this record and does so via the EIA-923 monthly survey form.

What I discovered is that it appears that EIA imputes monthly values for `A` data in the `gf` table, but not for data in the gen or bf tables. This means that when allocating data, the gf data is shaped, but the g and bf data is not.

The way that we currently handle this is that data in the g or bf data is allocated evenly to all months, which would mean that the imputed monthly data from the gf table would be allocated to each generator-month in proportion to the annual average generation or fuel consumption ratios. I'm think this may not necessarily be a bad approach, but open to other thoughts.

zaneselvans commented:
WTAF EIA?! They allocate for one of the tables but not the others? That's bonkers!

Is it clear how they're doing the imputation in the `generation_fuel` table? Can we apply the same approach in the `generation` and `boiler_fuel` tables?  Failing that it seems like ensuring that all three tables (which will be used in tandem for a bunch of different kinds of analyses) are doing the **same** thing (i.e. the flat distribution) even if that means losing some putative information from the monthly imputed values.

grgmiller commented:
I have a general idea of how they are doing the imputation, but not an exact formula to be able to re-create. When I've asked EIA about this in the past, they've pointed me to two documents, which include general information but no specifics: https://www.eia.gov/electricity/data/methodology/eiawebme.pdf and https://www.eia.gov/electricity/monthly/pdf/technotes.pdf. From what I can tell, they take monthly generation/fuel profiles from generators with similar PMs/fuels and use that to estimate the shape of annually-reported data. But I'm not sure how specifically they do this.

grgmiller commented:
So the most recent commit I added fixed an issue that was cropping up with older data. Merging `net_gen_alloc` and `fuel_alloc` together at the end of `allocate_gen_fuel_by_generator_energy_source()` requires unique keys in both data. However, I was discovering that in some years the allocation process was leading to duplicate values of `IDX_PM_ESC + ["generator_id", "energy_source_code_num"]`. However, it seems that the duplicate values, when grouped, still added up to the correct total, so to fix this, I added a groupby function before the merge to ensure unique keys in both datasets.

This does add a warning message about duplicate keys before grouping so that these can be investigated further.

grgmiller commented:
Ok question for @zaneselvans and @cmgosnell - I want to merge `dev` into this branch to prepare the PR for merging, but I have a bit of a conundrum. My forked branch that is the basis of this PR is also the branch of PUDL that the current version of OGE installs for its environment. However, updating OGE to work with the latest changes in the pudl dev branch is going to require a bit of work that I don't want to focus on right now. I think for the future, I've learned the lesson to not use the same branch to contribute to pudl that I use as a dependency for OGE. The immediate issue is that unless there is a way to update the branch that this PR is based on, I might need to close this PR, create a new branch that is separate from the branch that OGE depends on, and then open a new PR to merge in that branch. Are you aware of any ways around this?

grgmiller reviewed (commented):
Oops I had a huge backlog of comments that were pending! I keep forgetting that I have to actually submit my review
